File: Norgoth/apps/dashboard/src/lib/discord/message-compiler.ts
/** Authoring storage cap for TinyMCE-backed markdown (mirrors API). */
export const MAX_STORED_MARKDOWN_CHARS = 100_000;

export const DISCORD_DELIVERY_LIMITS = {
  content: 2000,
  embedDescription: 4096,
  embedTitle: 256,
  embedFooter: 2048,
  fieldName: 256,
  fieldValue: 1024,
  authorName: 256,
  embedFields: 25,
  total: 6000,
  maxEmbedsPerMessage: 10,
  maxMessagesPerDelivery: 5,
} as const;

export type CompileError = {
  code: string;
  message: string;
};

export type PreviewEmbed = {
  title?: string;
  description?: string;
  color?: number | string | null;
  footer?: string;
  footer_icon_url?: string;
  author?: { name?: string; url?: string; icon_url?: string };
  thumbnail_url?: string;
  image_url?: string;
  fields?: Array<{ name: string; value: string; inline?: boolean }>;
};

export type DiscordMessagePayload = {
  content?: string;
  embeds?: PreviewEmbed[];
};

export type CompileResult = {
  payloads: DiscordMessagePayload[];
  warnings: string[];
  errors: CompileError[];
};

function cappedLen(value: string | null | undefined, cap: number): number {
  const trimmed = value?.trim() ?? "";
  return trimmed ? Math.min(trimmed.length, cap) : 0;
}

function fieldsChars(
  fields: PreviewEmbed["fields"] | undefined,
): number {
  if (!fields?.length) return 0;
  return fields.slice(0, DISCORD_DELIVERY_LIMITS.embedFields).reduce((sum, field) => {
    return (
      sum +
      Math.min(field.name.length, DISCORD_DELIVERY_LIMITS.fieldName) +
      Math.min(field.value.length, DISCORD_DELIVERY_LIMITS.fieldValue)
    );
  }, 0);
}

function chromeChars(
  embed: PreviewEmbed,
  opts: {
    title: boolean;
    author: boolean;
    fields: boolean;
    footer: boolean;
  },
): number {
  let total = 0;
  if (opts.title) total += cappedLen(embed.title, DISCORD_DELIVERY_LIMITS.embedTitle);
  if (opts.author) total += cappedLen(embed.author?.name, DISCORD_DELIVERY_LIMITS.authorName);
  if (opts.fields) total += fieldsChars(embed.fields);
  if (opts.footer) total += cappedLen(embed.footer, DISCORD_DELIVERY_LIMITS.embedFooter);
  return total;
}

export function splitMarkdown(
  text: string,
  limit: number,
  firstLimit?: number,
): string[] {
  const trimmed = text.trim();
  if (!trimmed) return [];
  if (firstLimit === undefined && trimmed.length <= limit) return [trimmed];

  const chunks: string[] = [];
  let remaining = trimmed;
  let currentLimit = firstLimit ?? limit;

  while (remaining) {
    const thisLimit = currentLimit > 0 ? currentLimit : limit;
    currentLimit = limit;
    if (remaining.length <= thisLimit) {
      chunks.push(remaining.trim());
      break;
    }

    const window = remaining.slice(0, thisLimit + 1);
    let splitAt = -1;

    for (const marker of ["\n\n", "\n#", "\n-", "\n*", "\n1.", "\n"]) {
      const idx = window.lastIndexOf(marker);
      if (idx > thisLimit / 4) {
        splitAt = idx + marker.length;
        break;
      }
    }

    if (splitAt <= 0) splitAt = thisLimit;

    let candidate = remaining.slice(0, splitAt).trimEnd();
    if (!candidate) {
      candidate = remaining.slice(0, thisLimit);
      splitAt = thisLimit;
    }

    chunks.push(candidate);
    remaining = remaining.slice(splitAt).trimStart();
  }

  return chunks.filter(Boolean);
}

function previewEmbedChars(embed: PreviewEmbed): number {
  return (
    (embed.title?.length ?? 0) +
    (embed.description?.length ?? 0) +
    (embed.footer?.length ?? 0) +
    (embed.author?.name?.length ?? 0) +
    fieldsChars(embed.fields)
  );
}

function packEmbedGroups(cards: PreviewEmbed[]): PreviewEmbed[][] {
  const groups: PreviewEmbed[][] = [];
  let current: PreviewEmbed[] = [];
  let currentChars = 0;
  for (const card of cards) {
    const chars = previewEmbedChars(card);
    const overflowCount = current.length >= DISCORD_DELIVERY_LIMITS.maxEmbedsPerMessage;
    const overflowChars =
      current.length > 0 && currentChars + chars > DISCORD_DELIVERY_LIMITS.total;
    if (overflowCount || overflowChars) {
      groups.push(current);
      current = [];
      currentChars = 0;
    }
    current.push(card);
    currentChars += chars;
  }
  if (current.length) groups.push(current);
  return groups;
}

/** Preview-oriented compile: mirrors backend stacked-embed chunking. */
export function compileForPreview(
  content: string,
  embed?: PreviewEmbed | null,
): CompileResult {
  const body = content.trim();
  const source: PreviewEmbed = embed ?? {};
  const desc = source.description?.trim() ?? "";
  const hasChrome = Boolean(
    source.title?.trim() ||
      source.author?.name?.trim() ||
      (source.fields?.length ?? 0) > 0 ||
      source.footer?.trim() ||
      hasMedia(source.thumbnail_url) ||
      hasMedia(source.image_url),
  );

  const firstChrome = chromeChars(source, {
    title: true,
    author: true,
    fields: true,
    footer: false,
  });
  let firstDescBudget = Math.min(
    DISCORD_DELIVERY_LIMITS.embedDescription,
    Math.max(0, DISCORD_DELIVERY_LIMITS.total - firstChrome),
  );

  let descParts = desc ? splitMarkdown(desc, DISCORD_DELIVERY_LIMITS.embedDescription, firstDescBudget) : [];
  const footerChars = cappedLen(source.footer, DISCORD_DELIVERY_LIMITS.embedFooter);
  if (
    descParts.length === 1 &&
    footerChars &&
    firstChrome + descParts[0].length + footerChars > DISCORD_DELIVERY_LIMITS.total
  ) {
    firstDescBudget = Math.max(0, DISCORD_DELIVERY_LIMITS.total - firstChrome);
    descParts = splitMarkdown(desc, DISCORD_DELIVERY_LIMITS.embedDescription, firstDescBudget);
  }

  const contentParts = body ? splitMarkdown(body, DISCORD_DELIVERY_LIMITS.content) : [];

  if (!descParts.length && !contentParts.length && !hasChrome) {
    return { payloads: [], warnings: [], errors: [] };
  }

  const embedCards: PreviewEmbed[] = [];
  if (descParts.length || hasChrome) {
    const lastIndex = Math.max(descParts.length - 1, 0);
    const count = Math.max(descParts.length, hasChrome ? 1 : 0);
    for (let index = 0; index < count; index += 1) {
      const isFirst = index === 0;
      const isLast = index === lastIndex || (!descParts.length && isFirst);
      const partDesc = descParts[index]?.trim() ?? "";
      const card: PreviewEmbed = { color: source.color };
      if (partDesc) card.description = partDesc;
      if (isFirst) {
        if (source.title) card.title = source.title;
        if (source.author) card.author = source.author;
        if (source.thumbnail_url) card.thumbnail_url = source.thumbnail_url;
        if (source.image_url) card.image_url = source.image_url;
        if (source.fields) card.fields = source.fields;
      }
      if (isLast) {
        if (source.footer) card.footer = source.footer;
        if (source.footer_icon_url) card.footer_icon_url = source.footer_icon_url;
      }
      embedCards.push(card);
    }
  }

  const embedGroups: PreviewEmbed[][] = embedCards.length
    ? packEmbedGroups(embedCards)
    : [[]];

  const parts = contentParts.length ? [...contentParts] : [""];
  const segmentCount = Math.max(parts.length, embedGroups.length, 1);
  while (parts.length < segmentCount) parts.push("");
  while (embedGroups.length < segmentCount) embedGroups.push([]);

  if (segmentCount > DISCORD_DELIVERY_LIMITS.maxMessagesPerDelivery) {
    return {
      payloads: [],
      warnings: [],
      errors: [
        {
          code: "content_too_long_for_delivery",
          message: `Content requires ${segmentCount} messages but the limit is ${DISCORD_DELIVERY_LIMITS.maxMessagesPerDelivery}.`,
        },
      ],
    };
  }

  const payloads: DiscordMessagePayload[] = [];
  for (let index = 0; index < segmentCount; index += 1) {
    const payload: DiscordMessagePayload = {};
    const partContent = parts[index]?.trim();
    if (partContent) payload.content = partContent;
    if (embedGroups[index]?.length) payload.embeds = embedGroups[index];
    if (payload.content || payload.embeds?.length) payloads.push(payload);
  }

  return { payloads, warnings: [], errors: [] };
}

File: Norgoth/apps/api/app/services/discord/message_compiler.py
"""Compile authored markdown + embed JSON into Discord-safe message payloads."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from app.core.content_limits import (
    DISCORD_MAX_CONTENT,
    DISCORD_MAX_EMBED_DESCRIPTION,
    DISCORD_MAX_EMBED_FIELDS,
    DISCORD_MAX_EMBED_FOOTER,
    DISCORD_MAX_EMBED_TITLE,
    DISCORD_MAX_EMBED_TOTAL,
    DISCORD_MAX_EMBEDS_PER_MESSAGE,
    DISCORD_MAX_FIELD_NAME,
    DISCORD_MAX_FIELD_VALUE,
    DISCORD_MAX_MESSAGES_PER_DELIVERY,
)
from app.services.discord.embed_builder import DISCORD_LIMITS, build_embed_dict


@dataclass
class CompileError:
    code: str
    message: str


@dataclass
class CompileResult:
    payloads: list[dict[str, Any]] = field(default_factory=list)
    warnings: list[str] = field(default_factory=list)
    errors: list[CompileError] = field(default_factory=list)

    @property
    def ok(self) -> bool:
        return not self.errors and bool(self.payloads)


def _capped_len(value: Any, cap: int) -> int:
    if isinstance(value, str) and value.strip():
        return min(len(value.strip()), cap)
    return 0


def _fields_chars(fields: Any) -> int:
    if not isinstance(fields, list):
        return 0
    total = 0
    for raw in fields[:DISCORD_MAX_EMBED_FIELDS]:
        if not isinstance(raw, dict):
            continue
        total += min(len(str(raw.get("name") or "")), DISCORD_MAX_FIELD_NAME)
        total += min(len(str(raw.get("value") or "")), DISCORD_MAX_FIELD_VALUE)
    return total


def _author_chars(author: Any) -> int:
    if not isinstance(author, dict):
        return 0
    return _capped_len(author.get("name"), DISCORD_LIMITS["author_name"])


def _chrome_chars(
    embed_in: dict[str, Any],
    *,
    include_title: bool,
    include_author: bool,
    include_fields: bool,
    include_footer: bool,
) -> int:
    total = 0
    if include_title:
        total += _capped_len(embed_in.get("title"), DISCORD_MAX_EMBED_TITLE)
    if include_author:
        total += _author_chars(embed_in.get("author"))
    if include_fields:
        total += _fields_chars(embed_in.get("fields"))
    if include_footer:
        total += _capped_len(embed_in.get("footer"), DISCORD_MAX_EMBED_FOOTER)
    return total


def _split_markdown(text: str, limit: int, *, first_limit: int | None = None) -> list[str]:
    trimmed = text.strip()
    if not trimmed:
        return []
    if first_limit is None and len(trimmed) <= limit:
        return [trimmed]

    chunks: list[str] = []
    remaining = trimmed
    current_limit = first_limit if first_limit is not None else limit

    while remaining:
        this_limit = current_limit
        current_limit = limit
        if this_limit <= 0:
            this_limit = limit
        if len(remaining) <= this_limit:
            chunks.append(remaining.strip())
            break

        window = remaining[: this_limit + 1]
        split_at = -1

        for marker in ("\n\n", "\n#", "\n-", "\n*", "\n1.", "\n"):
            idx = window.rfind(marker)
            if idx > this_limit // 4:
                split_at = idx + len(marker)
                break

        if split_at <= 0:
            split_at = this_limit

        candidate = remaining[:split_at].rstrip()
        if not candidate:
            candidate = remaining[:this_limit]
            split_at = this_limit

        chunks.append(candidate)
        remaining = remaining[split_at:].lstrip()

    return [chunk for chunk in chunks if chunk.strip()]


def _split_embed_description(description: str, first_limit: int) -> list[str]:
    return _split_markdown(
        description,
        DISCORD_MAX_EMBED_DESCRIPTION,
        first_limit=max(first_limit, 0),
    )


def compile_discord_messages(
    *,
    content: str | None = None,
    embed_json: dict[str, Any] | None = None,
    fallback_name: str | None = None,
) -> CompileResult:
    """Turn stored markdown + embed into one or more Discord REST payloads.

    Long descriptions are split across stacked embeds (up to 10 per message)
    so community-rules posts can exceed a single 4096/6000 embed without
    truncating or failing compile.
    """

    result = CompileResult()
    body = (content or "").strip()
    embed_in = embed_json if isinstance(embed_json, dict) else {}

    description = embed_in.get("description")
    desc_text = description.strip() if isinstance(description, str) else ""
    title = embed_in.get("title")
    has_title = isinstance(title, str) and bool(title.strip())
    has_chrome = bool(
        has_title
        or _author_chars(embed_in.get("author"))
        or _fields_chars(embed_in.get("fields"))
        or _capped_len(embed_in.get("footer"), DISCORD_MAX_EMBED_FOOTER)
        or (isinstance(embed_in.get("thumbnail_url"), str) and embed_in.get("thumbnail_url", "").strip())
        or (isinstance(embed_in.get("image_url"), str) and embed_in.get("image_url", "").strip())
    )

    first_chrome = _chrome_chars(
        embed_in,
        include_title=True,
        include_author=True,
        include_fields=True,
        include_footer=False,
    )
    footer_chars = _capped_len(embed_in.get("footer"), DISCORD_MAX_EMBED_FOOTER)
    first_desc_budget = min(
        DISCORD_MAX_EMBED_DESCRIPTION,
        max(0, DISCORD_MAX_EMBED_TOTAL - first_chrome),
    )

    desc_parts = (
        _split_embed_description(desc_text, first_desc_budget) if desc_text else []
    )
    if (
        desc_parts
        and footer_chars
        and first_chrome + len(desc_parts[0]) + footer_chars > DISCORD_MAX_EMBED_TOTAL
        and len(desc_parts) == 1
    ):
        # Footer does not fit beside a maxed-out first description — continue.
        tighter = max(0, DISCORD_MAX_EMBED_TOTAL - first_chrome)
        desc_parts = _split_embed_description(desc_text, tighter)

    content_parts = _split_markdown(body, DISCORD_MAX_CONTENT) if body else []

    if not desc_parts and not content_parts and not has_chrome:
        built = build_embed_dict(embed_in)
        if built:
            result.payloads.append({"embeds": [built]})
            return result
        if fallback_name:
            result.payloads.append({"content": fallback_name[:DISCORD_MAX_CONTENT]})
            return result
        result.errors.append(
            CompileError(
                code="empty_message",
                message="Nothing to deliver.",
            )
        )
        return result

    embed_dicts: list[dict[str, Any]] = []
    if desc_parts or has_chrome:
        last_index = max(len(desc_parts) - 1, 0)
        for index in range(max(len(desc_parts), 1 if has_chrome else 0)):
            part_desc = desc_parts[index] if index < len(desc_parts) else ""
            is_first = index == 0
            is_last = index == last_index or (not desc_parts and is_first)
            segment = dict(embed_in)
            if part_desc:
                segment["description"] = part_desc
            else:
                segment.pop("description", None)
            if not is_first:
                segment.pop("title", None)
                segment.pop("author", None)
                segment.pop("thumbnail_url", None)
                segment.pop("image_url", None)
                segment.pop("fields", None)
            if not is_last:
                segment.pop("footer", None)
                segment.pop("footer_icon_url", None)
            built = build_embed_dict(segment)
            if not built:
                continue
            total = _built_total_chars(built)
            if total > DISCORD_MAX_EMBED_TOTAL:
                result.errors.append(
                    CompileError(
                        code="embed_total_exceeded",
                        message="Embed exceeds Discord's total character limit.",
                    )
                )
                return result
            embed_dicts.append(built)

    if not content_parts:
        content_parts = [""]
    while len(content_parts) > 1 and not content_parts[-1]:
        content_parts.pop()

    embed_groups: list[list[dict[str, Any]]] = []
    if embed_dicts:
        for start in range(0, len(embed_dicts), DISCORD_MAX_EMBEDS_PER_MESSAGE):
            embed_groups.append(embed_dicts[start : start + DISCORD_MAX_EMBEDS_PER_MESSAGE])
    else:
        embed_groups = [[]]

    segment_count = max(len(content_parts), len(embed_groups), 1)
    while len(content_parts) < segment_count:
        content_parts.append("")
    while len(embed_groups) < segment_count:
        embed_groups.append([])

    if segment_count > DISCORD_MAX_MESSAGES_PER_DELIVERY:
        result.errors.append(
            CompileError(
                code="content_too_long_for_delivery",
                message=(
                    f"Content requires {segment_count} messages but the limit is "
                    f"{DISCORD_MAX_MESSAGES_PER_DELIVERY}."
                ),
            )
        )
        return result

    for index in range(segment_count):
        payload: dict[str, Any] = {}
        part_content = content_parts[index].strip()
        if part_content:
            payload["content"] = part_content[:DISCORD_MAX_CONTENT]
        group = embed_groups[index]
        if group:
            payload["embeds"] = group
        if payload:
            result.payloads.append(payload)

    if not result.payloads:
        result.errors.append(
            CompileError(
                code="empty_message",
                message="Nothing to deliver.",
            )
        )

    return result


def _built_total_chars(built: dict[str, Any]) -> int:
    total = 0
    for key in ("title", "description"):
        value = built.get(key)
        if isinstance(value, str):
            total += len(value)
    footer = built.get("footer")
    if isinstance(footer, dict) and isinstance(footer.get("text"), str):
        total += len(footer["text"])
    author = built.get("author")
    if isinstance(author, dict) and isinstance(author.get("name"), str):
        total += len(author["name"])
    for field in built.get("fields") or []:
        if isinstance(field, dict):
            total += len(str(field.get("name") or ""))
            total += len(str(field.get("value") or ""))
    return total
